src/librustc/session/config.rs

```diff
@@ -109,6 +109,7 @@ impl OutputType {
             OutputType::Bitcode => "bc",
             OutputType::Assembly => "s",
             OutputType::LlvmAssembly => "ll",
+            OutputType::Object if cfg!(target_os = "windows") => "obj",
             OutputType::Object => "o",
```

Comment:
You may be able to clone `TargetOptions` (or whatever the type is called)? That or pre-compute the file names. cc @rust-lang/compiler